File: src/main/java/com/portofino/realtrainmodunofficial/item/TrainItem.java
package com.portofino.realtrainmodunofficial.item;

import com.portofino.realtrainmodunofficial.RealTrainModUnofficialComponents;
import com.portofino.realtrainmodunofficial.ClientHooks;
import com.portofino.realtrainmodunofficial.RealTrainModUnofficial;
import com.portofino.realtrainmodunofficial.blockentity.LargeRailCoreBlockEntity;
import com.portofino.realtrainmodunofficial.blockentity.RailCollisionBlockEntity;
import com.portofino.realtrainmodunofficial.entity.TrainEntity;
import com.portofino.realtrainmodunofficial.rail.util.RailMap;
import com.portofino.realtrainmodunofficial.vehicle.VehicleDefinition;
import com.portofino.realtrainmodunofficial.vehicle.VehicleRegistry;
import net.minecraft.util.Mth;
import net.minecraft.network.chat.Component;
import net.minecraft.core.BlockPos;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.Vec3;

import java.util.List;
import java.util.Locale;

public class TrainItem extends Item {
    public enum Category {
        ELECTRIC,
        DIESEL,
        TEST
    }

    private final Category category;

    public TrainItem() {
        this(Category.ELECTRIC);
    }

    public TrainItem(Category category) {
        super(new Properties());
        this.category = category == null ? Category.ELECTRIC : category;
    }

    /** スポーンクールダウン: 0.2秒 = 4 ticks */
    private static final int SPAWN_COOLDOWN_TICKS = 4;

    @Override
    public InteractionResult useOn(UseOnContext context) {
        Level level = context.getLevel();
        if (level.isClientSide()) {
            return InteractionResult.SUCCESS;
        }
        Player player = context.getPlayer();
        if (player == null) {
            return InteractionResult.PASS;
        }
        // クールダウン中はスポーン不可
        if (player.getCooldowns().isOnCooldown(this)) {
            return InteractionResult.PASS;
        }
        ItemStack stack = context.getItemInHand();
        String selectedId = stack.get(RealTrainModUnofficialComponents.SELECTED_MODEL_ID.get());
        VehicleDefinition def = VehicleRegistry.getById(selectedId);
        if (def == null || !accepts(def)) {
            def = VehicleRegistry.getAll().stream()
                .filter(this::accepts)
                .findFirst()
                .orElse(null);
        }
        if (def == null) {
            return InteractionResult.PASS;
        }
        Vec3 clickedPoint = context.getClickLocation();
        RailSpawnData spawnData = findNearestRailSpawn(level, context.getClickedPos(), clickedPoint, player.getYRot());
        if (spawnData == null) {
            player.displayClientMessage(Component.translatable("message.realtrainmodunofficial.train.must_be_on_rail"), true);
            return InteractionResult.FAIL;
        }
        if (isOccupiedSpawnArea(level, spawnData.x(), spawnData.y() + 0.25D, spawnData.z(), spawnData.yaw(), def)) {
            player.displayClientMessage(Component.translatable("message.realtrainmodunofficial.train.already_exists"), true);
            return InteractionResult.FAIL;
        }
        TrainEntity train = TrainEntity.create(level, def.getId(), spawnData.x(), spawnData.y(), spawnData.z(), spawnData.yaw(), def.getTrainDistance());
        if (train == null) {
            return InteractionResult.PASS;
        }
        train.initializeOnRail(spawnData.map(), spawnData.split(), spawnData.index());
        level.addFreshEntity(train);
        // クールダウン付与（サーバー側。クライアントにも自動同期される）
        player.getCooldowns().addCooldown(this, SPAWN_COOLDOWN_TICKS);
        return InteractionResult.sidedSuccess(false);
    }

    @Override
    public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand hand) {
        if (level.isClientSide) {
            ClientHooks.openTrainSelectScreen(player, player.getItemInHand(hand), category);
        }
        return InteractionResultHolder.success(player.getItemInHand(hand));
    }

    @Override
    public void appendHoverText(ItemStack stack, Item.TooltipContext context, List<Component> tooltip, TooltipFlag flag) {
        String selectedId = stack.get(RealTrainModUnofficialComponents.SELECTED_MODEL_ID.get());
        if (selectedId != null && !selectedId.isBlank()) {
            VehicleDefinition def = VehicleRegistry.getById(selectedId);
            String name = def != null ? def.getDisplayName() : selectedId;
            tooltip.add(Component.translatable("tooltip.realtrainmodunofficial.model.selected", name));
        } else {
            tooltip.add(Component.translatable("tooltip.realtrainmodunofficial.model.none"));
        }
    }

    public Category getCategory() {
        return category;
    }

    public boolean accepts(VehicleDefinition definition) {
        return accepts(category, definition);
    }

    public static boolean accepts(Category category, VehicleDefinition definition) {
        if (definition == null || definition.isCarType()) {
            return false;
        }
        String key = (safe(definition.getId()) + " " + safe(definition.getDisplayName()) + " "
            + safe(definition.getModelFile()) + " " + safe(definition.getVehicleType())).toLowerCase(java.util.Locale.ROOT);
        String type = safe(definition.getVehicleType()).toUpperCase(java.util.Locale.ROOT);
        boolean test = "TC".equals(type) || key.contains("test") || key.contains("shader") || key.contains("試験");
        return switch (category == null ? Category.ELECTRIC : category) {
            case TEST -> test;
            case DIESEL -> false;
            case ELECTRIC -> !test;
        };
    }

    private static String safe(String value) {
        return value == null ? "" : value;
    }

    private boolean isOccupiedSpawnArea(Level level, double x, double y, double z, float yaw, VehicleDefinition def) {
        double halfLength = Math.max(1.75D, def.getTrainDistance() * 0.5D);
        for (VehicleDefinition.BogieDefinition bogie : def.getBogies()) {
            halfLength = Math.max(halfLength, Math.abs(bogie.position().z) + 0.95D);
        }
        for (Vec3 seat : def.getAllSeatPositions()) {
            halfLength = Math.max(halfLength, Math.abs(seat.z) + 0.95D);
        }
        double halfWidth = getSpawnHalfWidth(def);
        double radius = Math.max(halfLength, halfWidth) + 1.0D;
        var bounds = new net.minecraft.world.phys.AABB(
            x - radius,
            y - 0.75D,
            z - radius,
            x + radius,
            y + 4.0D,
            z + radius
        );
        if (level instanceof net.minecraft.server.level.ServerLevel serverLevel) {
            TrainEntity.purgeDanglingTrainResidue(serverLevel, bounds);
        }
        final double finalHalfLength = halfLength;
        final double finalHalfWidth = halfWidth;
        List<TrainEntity> overlaps = level.getEntitiesOfClass(TrainEntity.class, bounds, entity -> entity != null && entity.isAlive() && !entity.isRemoved()).stream()
            .filter(entity -> Math.abs(entity.getY() - y) <= 3.5D)
            .filter(entity -> rectanglesOverlap(
                x, z, yaw, finalHalfWidth, finalHalfLength,
                entity.getX(), entity.getZ(), entity.getYRot(),
                entity.getBodyHalfWidthForPlacement() * 0.8D,
                entity.getBodyHalfLengthForPlacement() * 0.9D
            ))
            .toList();
        if (!overlaps.isEmpty()) {
            logSpawnOccupancy(level, "train_item", x, y, z, yaw, finalHalfWidth, finalHalfLength, overlaps);
            return true;
        }
        return false;
    }

    private static void logSpawnOccupancy(Level level, String source, double x, double y, double z, float yaw, double halfWidth, double halfLength, List<TrainEntity> overlaps) {
        StringBuilder entities = new StringBuilder();
        for (TrainEntity entity : overlaps) {
            if (entities.length() > 0) {
                entities.append(" | ");
            }
            entities.append(String.format(
                Locale.ROOT,
                "%s uuid=%s pos=(%.2f,%.2f,%.2f) yaw=%.1f half=(%.2f,%.2f) removed=%s alive=%s",
                entity.getVehicleId(),
                entity.getUUID(),
                entity.getX(),
                entity.getY(),
                entity.getZ(),
                entity.getYRot(),
                entity.getBodyHalfWidthForPlacement(),
                entity.getBodyHalfLengthForPlacement(),
                entity.isRemoved(),
                entity.isAlive()
            ));
        }
        RealTrainModUnofficial.LOGGER.warn(
            String.format(
                Locale.ROOT,
                "Spawn blocked [%s] level=%s spawn=(%.2f,%.2f,%.2f) yaw=%.1f half=(%.2f,%.2f) overlaps=%s",
                source,
                level.dimension().location(),
                x, y, z, yaw, halfWidth, halfLength,
                entities
            )
        );
    }

    private static boolean rectanglesOverlap(
        double ax, double az, float ayaw, double ahx, double ahz,
        double bx, double bz, float byaw, double bhx, double bhz
    ) {
        double[][] axes = new double[][]{
            axisFromYaw(ayaw),
            perpendicularAxisFromYaw(ayaw),
            axisFromYaw(byaw),
            perpendicularAxisFromYaw(byaw)
        };
        double dx = bx - ax;
        double dz = bz - az;
        for (double[] axis : axes) {
            double centerProjection = Math.abs(dx * axis[0] + dz * axis[1]);
            double aExtent = projectedExtent(ahx, ahz, ayaw, axis);
            double bExtent = projectedExtent(bhx, bhz, byaw, axis);
            if (centerProjection > aExtent + bExtent) {
                return false;
            }
        }
        return true;
    }

    private static double[] axisFromYaw(float yaw) {
        double yawRad = Math.toRadians(-yaw);
        return new double[]{Math.cos(yawRad), Math.sin(yawRad)};
    }

    private static double[] perpendicularAxisFromYaw(float yaw) {
        double[] axis = axisFromYaw(yaw);
        return new double[]{-axis[1], axis[0]};
    }

    private static double projectedExtent(double halfWidth, double halfLength, float yaw, double[] axis) {
        double[] forward = axisFromYaw(yaw);
        double[] side = perpendicularAxisFromYaw(yaw);
        return Math.abs(axis[0] * side[0] + axis[1] * side[1]) * halfWidth
            + Math.abs(axis[0] * forward[0] + axis[1] * forward[1]) * halfLength;
    }

    private static double getSpawnHalfWidth(VehicleDefinition def) {
        double halfWidth = 1.1D;
        for (Vec3 seat : def.getAllSeatPositions()) {
            halfWidth = Math.max(halfWidth, Math.abs(seat.x) + 0.3D);
        }
        for (VehicleDefinition.BogieDefinition bogie : def.getBogies()) {
            halfWidth = Math.max(halfWidth, Math.abs(bogie.position().x) + 0.45D);
        }
        return halfWidth;
    }

    private static RailSpawnData findNearestRailSpawn(Level level, BlockPos clickedPos, Vec3 clickedPoint, float preferredYaw) {
        RailMap clickedRailMap = getRailMapAt(level, clickedPos, clickedPoint);
        if (clickedRailMap != null) {
            return findNearestPointOnMap(clickedRailMap, clickedPoint, preferredYaw);
        }

        RailSpawnData best = null;
        double bestDistSq = Double.POSITIVE_INFINITY;
        int radius = 16;
        double cx = clickedPoint.x;
        double cy = clickedPoint.y;
        double cz = clickedPoint.z;

        for (int dx = -radius; dx <= radius; dx++) {
            for (int dy = -radius; dy <= radius; dy++) {
                for (int dz = -radius; dz <= radius; dz++) {
                    BlockPos p = clickedPos.offset(dx, dy, dz);
                    RailMap map = getRailMapAt(level, p, clickedPoint);
                    if (map == null) continue;
                    int max = getSpawnSplit(map);
                    for (int i = 0; i <= max; i++) {
                        double[] pos = map.getRailPos(max, i);
                        double x = pos[1];
                        double y = map.getRailHeight(max, i);
                        double z = pos[0];
                        double d2 = (x - cx) * (x - cx) + (y - cy) * (y - cy) + (z - cz) * (z - cz);
                        if (d2 < bestDistSq) {
                            bestDistSq = d2;
                            best = createSpawnData(map, max, i, preferredYaw);
                        }
                    }
                }
            }
        }
        // 吸着距離を約2ブロックに制限する。以前は 8 ブロック(64)圏内の最寄りレールを掴むため、
        // 隣の平行レールに列車が吸い付いていた (ユーザー報告)。クリックしたレール付近にのみ置く。
        return bestDistSq <= 4.0 ? best : null;
    }

    private static RailMap getRailMapAt(Level level, BlockPos pos, Vec3 targetPoint) {
        if (level.getBlockEntity(pos) instanceof LargeRailCoreBlockEntity core && core.isLoaded()) {
            return getNearestRailMap(core, targetPoint);
        }
        if (level.getBlockEntity(pos) instanceof RailCollisionBlockEntity collision) {
            BlockPos corePos = collision.getCorePos();
            if (corePos != null && level.getBlockEntity(corePos) instanceof LargeRailCoreBlockEntity core && core.isLoaded()) {
                return getNearestRailMap(core, targetPoint);
            }
        }
        // 道床ブロック(BallastBlock)もレールコアを保持する。道床はカーブ全長に敷かれるため、
        // レール中央の道床をクリックしても列車設置のレール判定が効く (ユーザー報告対応)。
        if (level.getBlockEntity(pos) instanceof com.portofino.realtrainmodunofficial.blockentity.BallastBlockEntity ballast) {
            BlockPos corePos = ballast.getCorePos();
            if (corePos != null && level.getBlockEntity(corePos) instanceof LargeRailCoreBlockEntity core && core.isLoaded()) {
                return getNearestRailMap(core, targetPoint);
            }
        }
        return null;
    }

    private static RailMap getNearestRailMap(LargeRailCoreBlockEntity core, Vec3 targetPoint) {
        RailMap[] maps = core.getAllRailMaps();
        if (maps.length == 0) return null;
        if (maps.length == 1) return maps[0];
        double cx = targetPoint.x;
        double cy = targetPoint.y;
        double cz = targetPoint.z;
        RailMap best = null;
        double bestDistSq = Double.POSITIVE_INFINITY;
        for (RailMap map : maps) {
            int max = getSpawnSplit(map);
            for (int i = 0; i <= max; i++) {
                double[] posData = map.getRailPos(max, i);
                double x = posData[1];
                double y = map.getRailHeight(max, i);
                double z = posData[0];
                double d2 = (x - cx) * (x - cx) + (y - cy) * (y - cy) + (z - cz) * (z - cz);
                if (d2 < bestDistSq) {
                    bestDistSq = d2;
                    best = map;
                }
            }
        }
        return best;
    }

    private static RailSpawnData findNearestPointOnMap(RailMap map, Vec3 clickedPoint, float preferredYaw) {
        if (map == null) return null;
        RailSpawnData best = null;
        double bestDistSq = Double.POSITIVE_INFINITY;
        int max = getSpawnSplit(map);
        double cx = clickedPoint.x;
        double cy = clickedPoint.y;
        double cz = clickedPoint.z;
        for (int i = 0; i <= max; i++) {
            double[] pos = map.getRailPos(max, i);
            double x = pos[1];
            double y = map.getRailHeight(max, i);
            double z = pos[0];
            double d2 = (x - cx) * (x - cx) + (y - cy) * (y - cy) + (z - cz) * (z - cz);
            if (d2 < bestDistSq) {
                bestDistSq = d2;
                best = createSpawnData(map, max, i, preferredYaw);
            }
        }
        return best;
    }

    private static RailSpawnData createSpawnData(RailMap map, int max, int index, float preferredYaw) {
        double[] p = map.getRailPos(max, index);
        double x = p[1];
        double y = map.getRailHeight(max, index);
        double z = p[0];
        float yaw = choosePreferredRailYaw(map.getRailYaw(max, index), preferredYaw);
        return new RailSpawnData(map, max, index, x, y, z, yaw);
    }

    private static float choosePreferredRailYaw(float railYaw, float preferredYaw) {
        float forwardDiff = Math.abs(Mth.wrapDegrees(railYaw - preferredYaw));
        float reverseYaw = Mth.wrapDegrees(railYaw + 180.0F);
        float reverseDiff = Math.abs(Mth.wrapDegrees(reverseYaw - preferredYaw));
        return reverseDiff < forwardDiff ? reverseYaw : railYaw;
    }

    private static int getSpawnSplit(RailMap map) {
        if (map == null) {
            return 64;
        }
        return Math.max(96, RailMap.curveSplitForLength(map.getHorizontalPathLength()) * 6);
    }

    private record RailSpawnData(RailMap map, int split, int index, double x, double y, double z, float yaw) {
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/item/TrainItem.kt
package cc.mirukuneko.realtrainmodrenewed.item

import cc.mirukuneko.realtrainmodrenewed.blockentity.BallastBlockEntity
import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedComponents
import cc.mirukuneko.realtrainmodrenewed.ClientHooks
import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.blockentity.LargeRailCoreBlockEntity
import cc.mirukuneko.realtrainmodrenewed.blockentity.RailCollisionBlockEntity
import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity
import cc.mirukuneko.realtrainmodrenewed.rail.util.RailMap
import cc.mirukuneko.realtrainmodrenewed.vehicle.VehicleDefinition
import cc.mirukuneko.realtrainmodrenewed.vehicle.VehicleRegistry
import net.minecraft.util.Mth
import net.minecraft.network.chat.Component
import net.minecraft.core.BlockPos
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.TooltipFlag
import net.minecraft.world.item.context.UseOnContext
import net.minecraft.world.level.ClipContext
import net.minecraft.world.level.Level
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.HitResult
import net.minecraft.world.phys.Vec3
import java.util.Locale
import java.util.function.Consumer
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin

class TrainItem : Item {
    enum class Category { ELECTRIC, DIESEL, TEST }

    @JvmField val category: Category

    companion object {
        private const val SPAWN_COOLDOWN_TICKS = 4
        private const val DEFAULT_PLACEMENT_HALF_WIDTH = 1.35
        private const val RAIL_CORE_SCAN_RADIUS_CHUNKS = 6
        private const val MAX_CLICK_TO_RAIL_DISTANCE_SQ = 64.0

        @JvmStatic
        fun accepts(category: Category?, definition: VehicleDefinition?): Boolean {
            if (definition == null || definition.isCarType()) return false
            val type = (definition.vehicleType ?: "").uppercase(Locale.ROOT)
            val test = type == "TEST"
            return when (category ?: Category.ELECTRIC) {
                Category.TEST -> test
                Category.DIESEL -> false
                Category.ELECTRIC -> !test
            }
        }

        fun rectanglesOverlap(ax: Double, az: Double, ayaw: Float, ahx: Double, ahz: Double, bx: Double, bz: Double, byaw: Float, bhx: Double, bhz: Double): Boolean {
            val axes = arrayOf(axisFromYaw(ayaw), perpendicularAxisFromYaw(ayaw), axisFromYaw(byaw), perpendicularAxisFromYaw(byaw))
            val dx = bx - ax; val dz = bz - az
            for (axis in axes) {
                if (abs(dx * axis[0] + dz * axis[1]) > projectedExtent(ahx, ahz, ayaw, axis) + projectedExtent(bhx, bhz, byaw, axis)) return false
            }
            return true
        }
        private fun axisFromYaw(yaw: Float): DoubleArray {
            val r = Math.toRadians(-yaw.toDouble())
            return doubleArrayOf(-sin(r), cos(r))
        }
        private fun perpendicularAxisFromYaw(yaw: Float): DoubleArray {
            val a = axisFromYaw(yaw); return doubleArrayOf(-a[1], a[0])
        }
        private fun projectedExtent(hw: Double, hl: Double, yaw: Float, axis: DoubleArray): Double {
            val fwd = axisFromYaw(yaw); val side = perpendicularAxisFromYaw(yaw)
            return abs(axis[0] * side[0] + axis[1] * side[1]) * hw + abs(axis[0] * fwd[0] + axis[1] * fwd[1]) * hl
        }
    }

    constructor() : this(Category.ELECTRIC)
    constructor(category: Category) : this(category, Properties())
    constructor(category: Category, properties: Properties) : super(properties) {
        this.category = category ?: Category.ELECTRIC
    }

    fun accepts(definition: VehicleDefinition): Boolean = accepts(category, definition)

    override fun useOn(context: UseOnContext): InteractionResult {
        val level = context.level
        val player = context.player ?: return InteractionResult.PASS
        if (level.isClientSide)
            return if (findNearestRailSpawn(level, context.clickedPos, context.clickLocation, player.yRot) != null)
                InteractionResult.SUCCESS else InteractionResult.PASS
        if (player.isShiftKeyDown) return InteractionResult.PASS
        if (player.cooldowns.isOnCooldown(context.itemInHand)) return InteractionResult.PASS

        val stack = context.itemInHand
        val selectedId = stack.get(RealTrainModRenewedComponents.SELECTED_MODEL_ID.get())
        var spawnVehicleId = selectedId
        var def = VehicleRegistry.getById(spawnVehicleId)
        if (def == null || !accepts(def))
            def = VehicleRegistry.getAll().firstOrNull { accepts(it) }?.also { spawnVehicleId = VehicleRegistry.getSelectionId(it) }
        if (def == null) return InteractionResult.PASS

        val spawnData = findNearestRailSpawn(level, context.clickedPos, context.clickLocation, player.yRot)
            ?: run { player.sendOverlayMessage(Component.translatable("message.realtrainmodrenewed.train.must_be_on_rail")); return InteractionResult.FAIL }

        if (isOccupiedSpawnArea(level, spawnData.x, spawnData.y + 0.25, spawnData.z, spawnData.yaw, def)) {
            player.sendOverlayMessage(Component.translatable("message.realtrainmodrenewed.train.already_exists"))
            return InteractionResult.FAIL
        }

        val train = TrainEntity.create(level, spawnVehicleId ?: def.id, spawnData.x, spawnData.y, spawnData.z, spawnData.yaw, def.trainDistance)
            ?: run {
                RealTrainModRenewed.LOGGER.warn("Train placement failed: could not create vehicle id={} display={} at ({}, {}, {})", spawnVehicleId ?: def.id, def.displayName, spawnData.x, spawnData.y, spawnData.z)
                player.sendOverlayMessage(Component.literal("列車を生成できませんでした(モデルIDを確認)"))
                return InteractionResult.FAIL
            }
        train.initializeOnRail(spawnData.map, spawnData.split, spawnData.index)
        level.addFreshEntity(train)
        player.cooldowns.addCooldown(context.itemInHand, SPAWN_COOLDOWN_TICKS)
        return InteractionResult.SUCCESS_SERVER
    }

    override fun use(level: Level, player: Player, hand: InteractionHand): InteractionResult {
        if (level.isClientSide && !isLookingAtBlock(level, player))
            ClientHooks.openTrainSelectScreen(player, player.getItemInHand(hand), category)
        return InteractionResult.SUCCESS
    }

    override fun appendHoverText(stack: ItemStack, context: TooltipContext, display: net.minecraft.world.item.component.TooltipDisplay, tooltip: Consumer<Component>, flag: TooltipFlag) {
        val selectedId = stack.get(RealTrainModRenewedComponents.SELECTED_MODEL_ID.get())
        if (!selectedId.isNullOrBlank()) {
            val def = VehicleRegistry.getById(selectedId)
            tooltip.accept(Component.translatable("tooltip.realtrainmodrenewed.model.selected", def?.displayName ?: selectedId))
        } else tooltip.accept(Component.translatable("tooltip.realtrainmodrenewed.model.none"))
    }

    private fun isLookingAtBlock(level: Level, player: Player): Boolean {
        val start = player.getEyePosition(1.0f)
        val end = start.add(player.getViewVector(1.0f).scale(player.blockInteractionRange().toDouble()))
        return level.clip(ClipContext(start, end, ClipContext.Block.OUTLINE, ClipContext.Fluid.NONE, player)).type == HitResult.Type.BLOCK
    }

    private fun isOccupiedSpawnArea(level: Level, x: Double, y: Double, z: Double, yaw: Float, def: VehicleDefinition): Boolean {
        val halfLength = placementHalfLength(def)
        val halfWidth = placementHalfWidth(def)
        var maxExistingHalfLength = halfLength
        var maxExistingHalfWidth = halfWidth
        for (knownDef in VehicleRegistry.getAll()) {
            maxExistingHalfLength = maxOf(maxExistingHalfLength, placementHalfLength(knownDef))
            maxExistingHalfWidth = maxOf(maxExistingHalfWidth, placementHalfWidth(knownDef))
        }
        val radius = halfLength + maxExistingHalfLength + maxExistingHalfWidth + 2.0
        val bounds = AABB(x - radius, y - 0.75, z - radius, x + radius, y + 4.0, z + radius)
        if (level is net.minecraft.server.level.ServerLevel) TrainEntity.purgeDanglingTrainResidue(level, bounds)
        val overlaps = level.getEntitiesOfClass(TrainEntity::class.java, bounds) { it.isAlive && !it.isRemoved }
            .filter { abs(it.y - y) <= 3.5 }
            .filter {
                rectanglesOverlap(
                    x, z, yaw, halfWidth, halfLength,
                    it.x, it.z, it.yRot, it.bodyHalfWidthForPlacement, it.bodyHalfLengthForPlacement
                )
            }
        if (overlaps.isNotEmpty()) {
            logSpawnOccupancy(level, "train_item", x, y, z, yaw, halfWidth, halfLength, overlaps)
            return true
        }
        return false
    }

    private fun placementHalfLength(def: VehicleDefinition): Double {
        var halfLength = maxOf(1.75, def.trainDistance.toDouble())
        for (bogie in def.bogies) halfLength = maxOf(halfLength, abs(bogie.position.z) + 0.95)
        return halfLength
    }

    private fun placementHalfWidth(def: VehicleDefinition): Double {
        var halfWidth = DEFAULT_PLACEMENT_HALF_WIDTH
        for (bogie in def.bogies) halfWidth = maxOf(halfWidth, abs(bogie.position.x) + 1.0)
        return halfWidth
    }

    private fun logSpawnOccupancy(level: Level, source: String, x: Double, y: Double, z: Double, yaw: Float, hw: Double, hl: Double, overlaps: List<TrainEntity>) {
        val entities = overlaps.joinToString(" | ") {
            String.format(Locale.ROOT, "%s uuid=%s pos=(%.2f,%.2f,%.2f) yaw=%.1f half=(%.2f,%.2f) removed=%s alive=%s",
                it.vehicleId, it.uuid, it.x, it.y, it.z, it.yRot, it.bodyHalfWidthForPlacement, it.bodyHalfLengthForPlacement, it.isRemoved, it.isAlive)
        }
        RealTrainModRenewed.LOGGER.warn("Spawn blocked [$source] level=${level.dimension().identifier()} spawn=($x,$y,$z) yaw=$yaw half=($hw,$hl) overlaps=$entities")
    }

    private fun findNearestRailSpawn(level: Level, clickedPos: BlockPos, clickedPoint: Vec3, preferredYaw: Float): RailSpawnData? {
        for (dy in -4..4) for (dx in -8..8) for (dz in -8..8) {
            val pos = clickedPos.offset(dx, dy, dz)
            val be = level.getBlockEntity(pos)
            when (be) {
                is LargeRailCoreBlockEntity -> if (be.isLoaded) {
                    val map = getNearestRailMap(be, clickedPoint)
                    if (map != null) return findNearestPointOnMap(map, clickedPoint, preferredYaw)
                }
                is RailCollisionBlockEntity -> {
                    val corePos = be.corePos ?: continue
                    val core = level.getBlockEntity(corePos)
                    if (core is LargeRailCoreBlockEntity && core.isLoaded) {
                        val map = getNearestRailMap(core, clickedPoint)
                        if (map != null) return findNearestPointOnMap(map, clickedPoint, preferredYaw)
                    }
                }
                is BallastBlockEntity -> {
                    val corePos = be.corePos ?: continue
                    val core = level.getBlockEntity(corePos)
                    if (core is LargeRailCoreBlockEntity && core.isLoaded) {
                        val map = getNearestRailMap(core, clickedPoint)
                        if (map != null) return findNearestPointOnMap(map, clickedPoint, preferredYaw)
                    }
                }
            }
        }
        return findNearestRailSpawnFromLoadedCores(level, clickedPos, clickedPoint, preferredYaw)
    }

    private fun findNearestRailSpawnFromLoadedCores(level: Level, clickedPos: BlockPos, clickedPoint: Vec3, preferredYaw: Float): RailSpawnData? {
        val centerChunkX = clickedPos.x shr 4
        val centerChunkZ = clickedPos.z shr 4
        var best: RailSpawnData? = null
        var bestDistSq = Double.POSITIVE_INFINITY
        for (cx in centerChunkX - RAIL_CORE_SCAN_RADIUS_CHUNKS..centerChunkX + RAIL_CORE_SCAN_RADIUS_CHUNKS) {
            for (cz in centerChunkZ - RAIL_CORE_SCAN_RADIUS_CHUNKS..centerChunkZ + RAIL_CORE_SCAN_RADIUS_CHUNKS) {
                val chunkCenter = BlockPos((cx shl 4) + 8, clickedPos.y, (cz shl 4) + 8)
                if (!level.hasChunkAt(chunkCenter)) continue
                val chunk = level.getChunk(cx, cz)
                for (pos in chunk.blockEntities.keys) {
                    val core = level.getBlockEntity(pos) as? LargeRailCoreBlockEntity ?: continue
                    if (!core.isLoaded) continue
                    for (map in core.allRailMaps) {
                        val spawn = findNearestPointOnMap(map, clickedPoint, preferredYaw) ?: continue
                        val distSq = spawn.distanceSq
                        if (distSq < bestDistSq) {
                            bestDistSq = distSq
                            best = spawn
                        }
                    }
                }
            }
        }
        return best?.takeIf { it.distanceSq <= MAX_CLICK_TO_RAIL_DISTANCE_SQ }
    }

    private fun getNearestRailMap(core: LargeRailCoreBlockEntity, targetPoint: Vec3): RailMap? {
        val maps = core.allRailMaps
        if (maps.isEmpty()) return null
        if (maps.size == 1) return maps[0]
        var best: RailMap? = null; var bestDistSq = Double.POSITIVE_INFINITY
        for (map in maps) {
            val max = getSpawnSplit(map)
            for (i in 0..max) {
                val p = map.getRailPos(max, i)
                val d2 = (p[1] - targetPoint.x).let { it * it } + (map.getRailHeight(max, i) - targetPoint.y).let { it * it } + (p[0] - targetPoint.z).let { it * it }
                if (d2 < bestDistSq) { bestDistSq = d2; best = map }
            }
        }
        return best
    }

    private fun findNearestPointOnMap(map: RailMap, clickedPoint: Vec3, preferredYaw: Float): RailSpawnData? {
        var best: RailSpawnData? = null; var bestDistSq = Double.POSITIVE_INFINITY
        val max = getSpawnSplit(map)
        for (i in 0..max) {
            val p = map.getRailPos(max, i)
            val d2 = (p[1] - clickedPoint.x).let { it * it } + (map.getRailHeight(max, i) - clickedPoint.y).let { it * it } + (p[0] - clickedPoint.z).let { it * it }
            if (d2 < bestDistSq) { bestDistSq = d2; best = RailSpawnData(map, max, i, p[1], map.getRailHeight(max, i), p[0], choosePreferredRailYaw(map.getRailYaw(max, i), preferredYaw), d2) }
        }
        return best
    }

    private fun choosePreferredRailYaw(railYaw: Float, preferredYaw: Float): Float {
        val fwdDiff = abs(Mth.wrapDegrees(railYaw - preferredYaw))
        val revYaw = Mth.wrapDegrees(railYaw + 180f)
        return if (abs(Mth.wrapDegrees(revYaw - preferredYaw)) < fwdDiff) revYaw else railYaw
    }

    private fun getSpawnSplit(map: RailMap?): Int {
        if (map == null) return 64
        return maxOf(96, RailMap.curveSplitForLength(map.getHorizontalPathLength()) * 6)
    }

    private data class RailSpawnData(val map: RailMap, val split: Int, val index: Int, val x: Double, val y: Double, val z: Double, val yaw: Float, val distanceSq: Double)
}
